vultr/resource_block_storage.go

```diff
@@ -0,0 +1,169 @@
+package vultr
+
+import (
+	"fmt"
+	"log"
+
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+func resourceBlockStorage() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceBlockStorageCreate,
+		Read:   resourceBlockStorageRead,
+		Update: resourceBlockStorageUpdate,
+		Delete: resourceBlockStorageDelete,
+
+		Schema: map[string]*schema.Schema{
+			"subid": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"label": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+
+			"dcid": {
+				Type:     schema.TypeInt,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"size_gb": {
+				Type:     schema.TypeInt,
+				Required: true,
+			},
+
+			"date_created": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"cost_per_month": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"status": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"attached_to_subid": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+		},
+	}
+}
+
+func resourceBlockStorageCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*Client)
+
+	label := d.Get("label").(string)
+	dcid := d.Get("dcid").(int)
+	sizeGb := d.Get("size_gb").(int)
+
+	log.Printf("[INFO] Creating new block storage")
+	storage, err := client.CreateBlockStorage(label, dcid, sizeGb)
+	if err != nil {
+		return fmt.Errorf("Error creating block storage: %v", err)
+	}
+	d.SetId(storage.ID)
+
+	old, new := d.GetChange("attached_to_subid")
+	if old != new {
+		if err := client.AttachBlockStorage(d.Id(), new.(string)); err != nil {
+			return fmt.Errorf("Error attaching block storage (%s) to %q: %v", d.Id(), new.(string), err)
+		}
+	}
+
+	return resourceBlockStorageRead(d, meta)
+}
+
+func resourceBlockStorageRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*Client)
+
+	storage, err := client.GetBlockStorage(d.Id())
+	if err != nil {
+		if err.Error() == "Invalid block storage." {
+			log.Printf("[WARN] Removing block storage (%s) because it is gone", d.Id())
+			d.SetId("")
+			return nil
+		}
+		d.SetId("")
```

Comment:
We should not set the id to “” in this case because there is no indication that the resource is gone. It may just be a transient error reading from the API. Plus, if we set this to “” we cannot log it in the next line.

Reply:
I was using it for debug and have forgotten to remove. Sorry :(